File: packages/database/src/util/dateRange.ts
import type { DateRange, QueryExpensesArgs } from '@mr-carson/shared-types';

function iso(d: Date): string {
  return d.toISOString().slice(0, 10);
}

function startOfWeek(d: Date): Date {
  // ISO week, Monday-start.
  const out = new Date(d);
  const day = (out.getUTCDay() + 6) % 7;
  out.setUTCDate(out.getUTCDate() - day);
  out.setUTCHours(0, 0, 0, 0);
  return out;
}

export function rangeFor(name: DateRange, now: Date = new Date()): { start: string; end: string } {
  const today = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));
  switch (name) {
    case 'today':
      return { start: iso(today), end: iso(today) };
    case 'yesterday': {
      const y = new Date(today);
      y.setUTCDate(y.getUTCDate() - 1);
      return { start: iso(y), end: iso(y) };
    }
    case 'this_week': {
      const start = startOfWeek(today);
      const end = new Date(start);
      end.setUTCDate(end.getUTCDate() + 6);
      return { start: iso(start), end: iso(end) };
    }
    case 'last_week': {
      const thisWeekStart = startOfWeek(today);
      const start = new Date(thisWeekStart);
      start.setUTCDate(start.getUTCDate() - 7);
      const end = new Date(thisWeekStart);
      end.setUTCDate(end.getUTCDate() - 1);
      return { start: iso(start), end: iso(end) };
    }
    case 'this_month': {
      const start = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), 1));
      const end = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth() + 1, 0));
      return { start: iso(start), end: iso(end) };
    }
    case 'last_month': {
      const start = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth() - 1, 1));
      const end = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), 0));
      return { start: iso(start), end: iso(end) };
    }
    case 'this_year': {
      const start = new Date(Date.UTC(today.getUTCFullYear(), 0, 1));
      const end = new Date(Date.UTC(today.getUTCFullYear(), 11, 31));
      return { start: iso(start), end: iso(end) };
    }
    case 'all_time':
      return { start: '0001-01-01', end: '9999-12-31' };
  }
}

export function resolveDateRange(args: QueryExpensesArgs): { start: string; end: string } | null {
  if (args.startDate && args.endDate) return { start: args.startDate, end: args.endDate };
  if (args.dateRange) return rangeFor(args.dateRange);
  return null;
}

File: apps/mobile/lib/domain/date_range.dart
import 'models/ai_models.dart';

// Ported from packages/database/src/util/dateRange.ts.
// ISO week, Monday-start. All math in UTC to match the TS source exactly.

class _Range {
  final String start;
  final String end;
  const _Range(this.start, this.end);
}

String _iso(DateTime d) => d.toIso8601String().substring(0, 10);

DateTime _startOfWeek(DateTime d) {
  // weekday: Mon=1..Sun=7  ->  days since Monday = weekday-1
  final daysSinceMonday = d.weekday - 1;
  return DateTime.utc(d.year, d.month, d.day - daysSinceMonday);
}

_Range _rangeFor(DateRange name, DateTime now) {
  final today = DateTime.utc(now.year, now.month, now.day);
  switch (name) {
    case DateRange.today:
      return _Range(_iso(today), _iso(today));
    case DateRange.yesterday:
      final y = today.subtract(const Duration(days: 1));
      return _Range(_iso(y), _iso(y));
    case DateRange.thisWeek:
      final start = _startOfWeek(today);
      return _Range(_iso(start), _iso(start.add(const Duration(days: 6))));
    case DateRange.lastWeek:
      final thisWeekStart = _startOfWeek(today);
      final start = thisWeekStart.subtract(const Duration(days: 7));
      final end = thisWeekStart.subtract(const Duration(days: 1));
      return _Range(_iso(start), _iso(end));
    case DateRange.thisMonth:
      final start = DateTime.utc(today.year, today.month, 1);
      final end = DateTime.utc(today.year, today.month + 1, 0); // day 0 = last of prev
      return _Range(_iso(start), _iso(end));
    case DateRange.lastMonth:
      final start = DateTime.utc(today.year, today.month - 1, 1);
      final end = DateTime.utc(today.year, today.month, 0);
      return _Range(_iso(start), _iso(end));
    case DateRange.thisYear:
      return _Range(_iso(DateTime.utc(today.year, 1, 1)),
          _iso(DateTime.utc(today.year, 12, 31)));
    case DateRange.allTime:
      return const _Range('0001-01-01', '9999-12-31');
  }
}

/// Returns null when no range is specified (caller leaves dates unbounded).
({String start, String end})? resolveDateRange(QueryExpensesArgs args,
    {DateTime? now}) {
  if (args.startDate != null && args.endDate != null) {
    return (start: args.startDate!, end: args.endDate!);
  }
  if (args.dateRange != null) {
    final r = _rangeFor(args.dateRange!, (now ?? DateTime.now()).toUtc());
    return (start: r.start, end: r.end);
  }
  return null;
}
